velox/functions/prestosql/Reduce.cpp

```diff
@@ -168,7 +178,12 @@ class ReduceFunction : public exec::VectorFunction {
           nullptr,
           &localResult);
     }
-    context.moveOrCopyResult(localResult, rows, result);
+    context.moveOrCopyResult(localResult, *nonNullRows, result);
+    for (int i = 0; i < flatArray->size(); i++) {
```

Comment:
this loop can be replaced with 
`result->addNulls(flatArray->rawNulls(), rows);`

Reply:
It doesn't work. Maybe because we passed nonNullRows to lambda and can't apply flatArray `rawNulls` here.

Reply:
@yma11 in that case, you can do 

```
partialResult->addNulls(flatArray->rawNulls(), rows);
context.moveOrCopyResult(localResult, rows, result);
```

since partialResult has already been initialized with rows.end() size.

Reply:
Tried but some TCs failed. Besides it doesn't always make sense to apply `addNulls` on `partialResult`. Take `coalesce(s, 0)` for example, the null value will be incorrectly updated to `0` in this case.

Reply:
@yma11 
> Tried but some TCs failed. 

Curious which test cases failed and why? I might be missing something here so would love to know.

> Besides it doesn't always make sense to apply addNulls on partialResult. Take coalesce(s, 0) for example, the null value will be eventually incorrectly updated to 0 in this case.

The nulls we are adding here is in the last step and are the nulls rows in the array which should not be passed to the lambda. these should then we copied over directly to the final result in moveOrCopyResult() and should end up as null. Curious whats causing the result to be overwritten in the coalesce example you mentioned.

Reply:
> Curious which test cases failed and why? I might be missing something here so would love to know.
> 
The existing TCs `elementIndicesOverwrite` and `try` fail with errors like: 
```
Value of: expected->equalValueAt(actual.get(), i, i)
  Actual: false
Expected: true
at 0: expected -2, but got null
WARNING: Logging before InitGoogleLogging() is written to STDERR
E0906 15:44:39.674255 419537 Exceptions.h:69] Line: ../.././velox/vector/ConstantVector.h:276, Function:addNulls, Expression:  addNulls not supported, Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: addNulls not supported
Retriable: False
```
> The nulls we are adding here is in the last step and are the nulls rows in the array which should not be passed to the lambda. these should then we copied over directly to the final result in moveOrCopyResult() and should end up as null. Curious whats causing the result to be overwritten in the coalesce example you mentioned.
```
partialResult->addNulls(flatArray->rawNulls(), rows);
context.moveOrCopyResult(localResult, rows, result);
```
Above lines will guarantee `partialResult` will have correct NULL values. But I think it doesn't distinguish it's a null value or actually a Null array. Take UT in this PR as example, `reduce(c0, 0, (s, x) -> s + x, s -> coalesce(s, 0) * 10)`, partialResult will be null for both an array like `(1, 2, null, value)` and NULL array. And in the finish function ` coalesce(s, 0)`, the null in `partialResult` will become 0 not matter what it represents.  This is what this PR wants to address.

Reply:
@yma11 I looked into this issue and noticed an unexpected pre-existing API implementation + a bug in this PR

(also to clarify, I mistakingly mentioned using `partialResult` when adding nulls, instead of using `localResult`)

**1. unexpected pre-existing API implementation:**

I was curious why BaseVector::addNulls is not working properly to I dug into its implementation. Turns out if `flatArray->rawNulls()` is nullptr then it treats the unselected rows in `rows` as nulls and adds those nulls to the vector. This is not clear from the method comment. I'll look into where else this might be causing bugs later. The expectation here is that when using this API, we should add all the nulls in `flatArray->rawNulls()` for rows that are active in `rows`. So, to ensure we do that, please use this instead:

```
   if (flatArray->rawNulls()) {
      localResult->addNulls(flatArray->rawNulls(), rows);
    }
    context.moveOrCopyResult(localResult, rows, result);
```

**2. a bug in this PR**

if we proceed with the current implementation of using this:
```
   for (int i = 0; i < flatArray->size(); i++) {
      if (flatArray->isNullAt(i)) {
        result->setNull(i, true);
      }
    }
```
This would touch rows that are not marked as "active" in `rows` which will cause incorrect results as those rows might be filled in by other branches of the expression tree. This bug was exposed by investigating failure in `TEST_F(ReduceTest, finalSelection)` (see comment there on how to fix the test after applying the suggested implementation as suggested above.

Reply:
Thanks for finding this out. The `addNulls` works as you suggested. Please take a look again.